src/ir_passes/mono/monodeferred.rs

```diff
@@ -314,6 +326,45 @@ impl MonoDeferred<'_, '_> {
         self.monosig.base.extend_cmds(body);
     }
 
+    fn fact(&mut self, fact: &ir::Fact) -> Option<ir::Fact> {
+        let ir::Fact { prop, reason, .. } = fact;
+
+        let prop = prop.ul();
+        log::debug!("Fact: {}", self.underlying.display(prop));
+        let (params, events) = prop.idx().relevant_vars(self.underlying.comp());
+        assert!(events.is_empty(), "Facts cannot mention events.");
+
+        if !params
+            .into_iter()
+            .map(|p| p.ul())
+            .all(|idx| self.monosig.binding.get(&idx).is_some())
+        {
+            // Discards the assertion if it references parameters that can't be resolved (bundle parameters, etc)
+            // TODO(edmund): Find a better solution to this - we should resolve bundle assertions when bundles are unrolled.
+            None
+        } else {
+            log::debug!(
+                "Resolving: {} with binding {}",
+                self.underlying.display(prop),
+                self.monosig.binding_rep(&self.underlying)
+            );
+            let prop = self.prop(prop);
+            log::debug!("Resolved: {}", self.monosig.base.display(prop));
+            let prop = self
+                .monosig
+                .base
+                .resolve_prop(self.monosig.base.get(prop).clone())
+                .get();
+
+            let reason = reason.ul();
+            let reason =
+                self.monosig.info(&self.underlying, self.pass, reason).get();
+
+            // After monomorphization, both assumes and asserts become asserts.
+            Some(ir::Fact::assert(prop, reason))
```

Comment:
We should a “generated” variant to ir::Reason which itself contains a Reason. Then, in diagnostics reporting, we can accumulate all the asserts that are generated from one source level assert and say which assert or assume caused the failure.